Translate DOM Elements (#45)
reactjs/pt-BR.reactjs.org: content/docs/reference-dom-elements.md

  - "tips/dangerously-set-inner-html.html"
---

> O React implementa um sistema DOM independente ao browser visando performance e compatibilidade entre browsers. Aproveitamos a oportunidade para melhorar algumas implementações do DOM no browser.

Reviewer: ```suggestion
O React implementa um sistema DOM independente ao navegador visando performance e compatibilidade entre os navegadores. Aproveitamos a oportunidade para melhorar algumas implementações do DOM no navegador.
```

Author: Todas as ocorrências de 'browser' foram traduzidas para 'navegador'

> No React, todas as propriedades e atributos (incluindo manipuladores de eventos) devem estar em camelCase. Por exemplo, o atributo HTML `tabindex` corresponde ao atributo `tabIndex` no React. As exceções são os atributos `aria-*` e `data-*`, os quais devem estar em lowercase. Por exemplo, você pode manter `aria-label` como `aria-label`.

Reviewer: ```suggestion
No React, todas as propriedades e atributos (incluindo manipuladores de eventos) devem estar em _camelCase_. Por exemplo, o atributo HTML `tabindex` corresponde ao atributo `tabIndex` no React. As exceções são os atributos `aria-*` e `data-*`, os quais devem estar em _lowercase_. Por exemplo, você pode manter `aria-label` como `aria-label`.
```

## Diferenças nos Atributos {#differences-in-attributes}

Existem alguns atributos que funcionam de formas diferentes no React e no HTML:

### checked {#checked}

> O atributo `checked` é suportado por componentes `<input>` do 'type' `checkbox` ou `radio `. Você pode usa-lo para definir se o componente está checado. Isso é útil para construção de componentes controlados(controlled components). `defaultChecked` é o equivalente para componentes não controlados(uncontrolled components), que define se o componente está checado na primeira vez em que é montado.

Reviewer 2: No dia a dia acabamos usando `checado` não consigo pensar em uma palavra melhor... 
Seguimos com a palavra `checado` mesmo?

> O atributo `checked` é suportado por componentes `<input>` do 'type' `checkbox` ou `radio `. Você pode usa-lo para definir se o componente está checado. Isso é útil para construção de componentes controlados(controlled components). `defaultChecked` é o equivalente para componentes não controlados(uncontrolled components), que define se o componente está checado na primeira vez em que é montado.

Reviewer: ```suggestion
O atributo `checked` é suportado por componentes `<input>` do tipo (_`type`_) `checkbox` ou `radio `. Você pode usá-lo para definir se o componente está checado. Isso é útil para construção de componentes controlados (_controlled components_). `defaultChecked` é o equivalente para componentes não controlados (_uncontrolled components_), que define se o componente está checado na primeira vez em que é montado.
```

### className {#classname}

Para especificar uma classe CSS, utilize o atributo `className`. Isso se aplica para todos os DOM e elementos SVG regulares como `<div>`, `<a>`, e outros.

> Se você usar React com Web Components (prática incomum), use o atributo `class`.

Reviewer: ```suggestion
Se você usar React com Web Components (que é uma prática incomum), use o atributo `class`.
```

### dangerouslySetInnerHTML {#dangerouslysetinnerhtml}

> `dangerouslySetInnerHTML` é a substituição do React para o uso do `innerHTML` no DOM do browser. Em geral, setar HTML através de código é arriscado pois é fácil expor inadvertidamente seus usuários a um ataque de [cross-site scripting (XSS)](https://en.wikipedia.org/wiki/Cross-site_scripting). Então, você pode definir HTML diretamente através do React, mas vocẽ tem que digitar `dangerouslySetInnerHTML` e passar um objeto com a chave `__html`, para te relembrar que é perigoso. Por exemplo: -->

Reviewer 2: ```suggestion
`dangerouslySetInnerHTML` é a substituição do React para o uso do `innerHTML` no DOM do navegador. Em geral, adicionar HTML através de código é arriscado pois é fácil expor inadvertidamente seus usuários a um ataque de [cross-site scripting (XSS)](https://pt.wikipedia.org/wiki/Cross-site_scripting). Então, você pode definir HTML diretamente através do React, mas vocẽ tem que digitar `dangerouslySetInnerHTML` e passar um objeto com a chave `__html`, para te relembrar que é perigoso. Por exemplo: -->
```

Reviewer 2: - browser pode ser traduzido.
- usar o termo setar é comun, mas acredito que podemos evitar

> `dangerouslySetInnerHTML` é a substituição do React para o uso do `innerHTML` no DOM do navegador. Em geral, adicionar HTML através de código é arriscado pois é fácil expor inadvertidamente seus usuários a um ataque de [cross-site scripting (XSS)](https://pt.wikipedia.org/wiki/Cross-site_scripting). Então, você pode definir HTML diretamente através do React, mas vocẽ tem que digitar `dangerouslySetInnerHTML` e passar um objeto com a chave `__html`, para te relembrar que é perigoso. Por exemplo: -->

Reviewer: ```suggestion
`dangerouslySetInnerHTML` é o substituto do React para o uso do `innerHTML` no DOM do navegador. Em geral, adicionar HTML através de código é arriscado pois é fácil expor inadvertidamente seus usuários a um ataque de [cross-site scripting (XSS)](https://pt.wikipedia.org/wiki/Cross-site_scripting). Então, você pode definir HTML diretamente através do React, mas vocẽ tem que digitar `dangerouslySetInnerHTML` e passar um objeto com a chave `__html`, para te relembrar que é perigoso. Por exemplo: -->
```

### htmlFor {#htmlfor}

> Já que `for` é uma palavra reservada em JavaScript, elementos em React usam `htmlFor` no lugar.

Reviewer: ```suggestion
Já que `for` é uma palavra reservada em JavaScript, em vez disso elementos React usam `htmlFor`.
```

### onChange {#onchange}

> O evento `onChange` se comporta como você esperaria que ele se comportasse: sempre que um campo do formulário muda, este evento é disparado. Nós intencionalmente não utilizamos o comportamento existente do browser porque `onChange` é um termo inadequado para seu comportamento e o React depende deste evento para manipular input do usuário em tempo real.

Reviewer: ```suggestion
O evento `onChange` se comporta como você esperaria que ele se comportasse: sempre que um campo do formulário muda, este evento é disparado. Nós intencionalmente não utilizamos o comportamento existente do navegador porque `onChange` é um termo inadequado para seu comportamento e o React depende deste evento para manipular input do usuário em tempo real.
```

### selected {#selected}

> O atributo `selected` é suportado pelos componentes `<option>`. Você pode usa-lo para definir se um componente está selecionado. Isso é útil para construir componentes controlados(controlled components).

Reviewer: ```suggestion
O atributo `selected` é suportado pelos componentes `<option>`. Você pode usá-lo para definir se um componente está selecionado. Isso é útil para construir componentes controlados (_controlled components_).
```

### style {#style}

>Nota
>

> Alguns exemplos na documentação que utilizam `style` por conveniência, mas que ao usar o atributo `style` como a forma principal de estilizar elementos geralmente não é recomendado.** Em alguns casos, [`className`](#classname) deve ser usado para referenciar classes definidas em um arquivo de estilo CSS externo. `style` é geralmente usado em aplicações React para adicionar estilos computados dinamicamente em tempo de renderização. Veja também[FAQ: Styling and CSS](/docs/faq-styling.html).

Reviewer 2: ```suggestion
Alguns exemplos na documentação utilizam `style` por conveniência, mas **usar o atributo `style` como a forma principal de estilizar elementos geralmente não é recomendado.** Na maioria dos casos, [`className`](#classname) deve ser usado para referenciar classes definidas em um arquivo de estilo CSS externo. `style` é geralmente usado em aplicações React para adicionar estilos computados dinamicamente em tempo de renderização. Veja também [FAQ: Styling and CSS](/docs/faq-styling.html).
```

Reviewer 2: O termo `FAQ: Styling and CSS` talvez deva ser traduzido também, porém o link NÃO deve ser traduzido.

> O atributo `style` aceita um objeto JavaScript com propriedades em camelCase ao invés de uma string CSS. Isso é consistente com a propriedade do Javascript `style` 

Reviewer: ```suggestion
O atributo `style` aceita um objeto JavaScript com propriedades em _camelCase_ ao invés de uma string CSS. Isso é consistente com a propriedade do JavaScript `style` 
```

```js
// Style resultante: '10px'
<div style={{ height: 10 }}>
  Hello World!
</div>

// Style resultante: '10%'
<div style={{ height: '10%' }}>
  Hello World!
</div>
```

Nem todas as propriedades de style são convertidas para strings pixel. Algumas permanecem sem unidade (exemplo: `zoom`, `order`, `flex`). Uma lista completa com as propriedades sem unidade pode ser vista [aqui](https://github.com/facebook/react/blob/4131af3e4bf52f3a003537ec95a1655147c81270/src/renderers/dom/shared/CSSProperty.js#L15-L59).

### suppressContentEditableWarning {#suppresscontenteditablewarning}

Normalmente, existe uma advertência quando um elemento que contém outros elementos também está marcado como `contentEditable`, porque não funcionará. Esse atributo suprime essa advertência. Não o use, a não ser que você esteja construindo uma biblioteca como a [Draft.js](https://facebook.github.io/draft-js/) que gerencia `contentEditable` manualmente.

### suppressHydrationWarning {#suppresshydrationwarning}

> Se você usa renderização React no lado do server, normalmente existe uma advertência para quando o server e o client renderizam conteúdo de formas diferentes. Entretanto, em alguns casos raros, é muito difícil ou impossível de garantir a correspondência exata. Por exemplo, é esperado que timestamps estejam diferentes no server e no client. 

Reviewer 3: ```suggestion
Se você usa a renderização do React no lado do servidor (server-side), normalmente existe um alerta para quando o servidor (server) e o lado do cliente (client) renderizam conteúdo de formas diferentes. Entretanto, em alguns casos raros, é muito difícil ou impossível de garantir a correspondência exata. Por exemplo, é esperado que timestamps estejam diferentes no servidor (server) e no cliente (client). 

```

> Se você definir `suppressHydrationWarning` para `true`, o React irá te advertir sobre incompatibilidades nos atributos e no conteúdo daquele elemento. Isto só funciona dentro de 1 nível de profundidade, e é pretendido ser usado como uma saída de emergência. Não use excessivamente. Você pode ler mais sobre hydration na [documentação do `ReactDOM.hydrate()`](/docs/react-dom.html#hydrate).

Reviewer 3: ```suggestion
Se você definir `suppressHydrationWarning` para `true`, o React te advertirá sobre incompatibilidades nos atributos e no conteúdo daquele elemento. Isto só funciona dentro de 1 nível de profundidade e é será usado como uma saída de emergência. Não use excessivamente. Você pode ler mais sobre hydration na [documentação do `ReactDOM.hydrate()`](/docs/react-dom.html#hydrate).
```

### value {#value}

O atributo `value` é suportado pelos componentes `<input>` e `<textarea>`. Você pode usa-lo para definir o valor do componente. Isso é útil para construção de componentes controlados(controlled components). `defaultValue` é o equivalente para componentes não controlados(uncontrolled components), que define o valor do componente quando este é montado pela primeira vez.

## Todos os atributos HTML suportados {#todos-os-atributos-html-suportados}

A partir do React 16, qualquer atributo padrão [ou customizado](/blog/2017/09/08/dom-attributes-in-react-16.html) são integralmente suportados.

> O React sempre forneceu uma API centrada a JavaScript para o DOM. Uma vez que os componentes React geralmente recebem tanto props customizadas quanto props relacionadas ao DOM. O React usa a convenção `camelCase` assim como as APIs do DOM:

Reviewer 3: ```suggestion
O React sempre forneceu uma API centrada no JavaScript para o DOM. Uma vez que os componentes React geralmente recebem tanto props customizadas quanto props relacionadas ao DOM. O React usa a convenção `camelCase` assim como as APIs do DOM:
```

}
```

> Perceba que estes estilos não são auto prefixados. Para serem compatíveis com browsers antigos você precisa fornecer as propriedades de estilos correspondentes:

Reviewer 2: ```suggestion
Perceba que estes estilos não são auto prefixados. Para serem compatíveis com navegadores antigos você precisa fornecer as propriedades de estilos correspondentes:
```

```js
const divStyle = {
  WebkitTransition: 'all', // perceba o 'W' maiúsculo aqui

>   msTransition: 'all' // 'ms' é o único prefixo vendor minúsculo

Reviewer: ```suggestion
  msTransition: 'all' // 'ms' é o único prefixo de fornecedor minúsculo
```

};

function ComponentWithTransition() {

>   return <div style={divStyle}>Isto deve funcionar em diferentes browsers</div>;

Reviewer 2: ```suggestion
  return <div style={divStyle}>Isto deve funcionar em diferentes navegadores</div>;
```

}
```

> chaves de style são em camelCase com o intuito de serem consistentes com o acesso de propriedades que são nós do DOM através do JS  (ex. `node.style.backgroundImage`). prefixos Vendor [diferentes de `ms`](http://www.andismith.com/blog/2012/02/modernizr-prefixed/) devem começar com a letra maiúscula. É por isso que `WebkitTransition` tem um "W" maiúsculo.

Reviewer: ```suggestion
Chaves de _style_ são em _camelCase_ com o intuito de serem consistentes com o acesso de propriedades que são nós do DOM através do JS  (ex. `node.style.backgroundImage`). Prefixos de fornecedor (_vendor prefixes_) [diferentes de `ms`](http://www.andismith.com/blog/2012/02/modernizr-prefixed/) devem começar com a letra maiúscula. É por isso que `WebkitTransition` tem um "W" maiúsculo.
```

```js
// Style resultante: '10px'
<div style={{ height: 10 }}>
  Hello World!
</div>

// Style resultante: '10%'
<div style={{ height: '10%' }}>
  Hello World!
</div>
```

Nem todas as propriedades de estilos são convertidas para strings pixel. Algumas permanecem sem unidade (exemplo: `zoom`, `order`, `flex`). Uma lista completa com as propriedades sem unidade pode ser vista [aqui](https://github.com/facebook/react/blob/4131af3e4bf52f3a003537ec95a1655147c81270/src/renderers/dom/shared/CSSProperty.js#L15-L59)

### suppressContentEditableWarning {#suppresscontenteditablewarning}

Normalmente, existe um alerta quando um elemento contém outros elementos que também estão marcados como `contentEditable`. Assim sendo, não funcionará. Esse atributo interrompe esse alerta. Não faça uso dele. A não ser que você esteja construindo uma biblioteca como a [Draft.js](https://facebook.github.io/draft-js/) que gerencia `contentEditable` manualmente.

### suppressHydrationWarning {#suppresshydrationwarning}

> Se você usa a renderização do React no lado do servidor (server-side), normalmente existe um alerta para quando o servidor (server) e o lado do cliente (client) renderizam conteúdo de formas diferentes. Entretanto, em alguns casos raros, é muito difícil ou impossível de garantir a correspondência exata. Por exemplo, é esperado que timestamps estejam diferentes no servidor (server) e no cliente (client). 

Reviewer: ```suggestion
Se você usa a renderização do React no lado do servidor, normalmente existe um alerta para quando o servidor e o lado do cliente renderizam conteúdo de formas diferentes. Entretanto, em alguns casos raros, é muito difícil ou impossível de garantir a correspondência exata. Por exemplo, é esperado que _timestamps_ estejam diferentes no servidor e no cliente. 
```

> Se você definir `suppressHydrationWarning` para `true`, o React te advertirá sobre incompatibilidades nos atributos e no conteúdo daquele elemento. Isto só funciona dentro de 1 nível de profundidade e é será usado como uma saída de emergência. Não use excessivamente. Você pode ler mais sobre hydration na [documentação do `ReactDOM.hydrate()`](/docs/react-dom.html#hydrate).

Reviewer: ```suggestion
Se você definir `suppressHydrationWarning` para `true`, o React te avisará sobre incompatibilidades nos atributos e no conteúdo daquele elemento. Isto só funciona dentro de 1 nível de profundidade e destina-se a ser usado como uma saída de emergência. Não use excessivamente. Você pode ler mais sobre _hydration_ na [documentação do `ReactDOM.hydrate()`](/docs/react-dom.html#hydrate).
```

### value {#value}

> O atributo `value` é suportado pelos componentes `<input>` e `<textarea>`. Você pode usa-lo para definir o valor do componente. Isso é útil para construção de componentes controlados(controlled components). `defaultValue` é o equivalente para componentes não controlados(uncontrolled components), que define o valor do componente quando este é montado pela primeira vez.

Reviewer: ```suggestion
O atributo `value` é suportado pelos componentes `<input>` e `<textarea>`. Você pode usá-lo para definir o valor do componente. Isso é útil para construção de componentes controlados (_controlled components_). `defaultValue` é o equivalente para componentes não controlados (_uncontrolled components_), que define o valor do componente quando este é montado pela primeira vez.
```

> ## Todos os atributos HTML suportados {#todos-os-atributos-html-suportados}

Reviewer 2: ```suggestion
## Todos os atributos HTML suportados {#all-supported-html-attributes}
```

Reviewer 2: Não devemos traduzir as ancoras

> A partir do React 16, qualquer atributo padrão [ou personalizado](/blog/2017/09/08/dom-attributes-in-react-16.html) são integralmente suportados.

Reviewer: ```suggestion
A partir do React 16, qualquer atributo padrão [ou personalizado](/blog/2017/09/08/dom-attributes-in-react-16.html) é totalmente suportado.
```

> O React sempre forneceu uma API centrada no JavaScript para o DOM. Uma vez que os componentes React geralmente recebem tanto props customizadas quanto props relacionadas ao DOM. O React usa a convenção `camelCase` assim como as APIs do DOM:

Reviewer 2: customizadas é OK para mim. Mas num seria melhor colocar o termo em portugues? Personalizada ou algo do tipo

Author: Acredito que personalizada seja a melhor tradução mesmo. Feito.

> O React sempre forneceu uma API centrada no JavaScript para o DOM. Uma vez que os componentes React geralmente recebem tanto props personalizadas quanto props relacionadas ao DOM. O React usa a convenção `camelCase` assim como as APIs do DOM:

Reviewer: ```suggestion
O React sempre forneceu uma API centrada no JavaScript para o DOM. Uma vez que os componentes React geralmente recebem tanto props personalizadas quanto props relacionadas ao DOM, o React usa a convenção `camelCase` assim como as APIs do DOM:
```

tabIndex target title type useMap value width wmode wrap
```

> Semelhantemente, todos os atributos SVG são totalmente suportados:

Reviewer 3: ```suggestion
Igualmente, todos os atributos SVG são totalmente suportados:
```

contentStyleType cursor cx cy d decelerate descent diffuseConstant direction
display divisor dominantBaseline dur dx dy edgeMode elevation enableBackground
end exponent externalResourcesRequired fill fillOpacity fillRule filter

> filterRes filterUnits floodColor floodOpacity focusable fontFamily fontSizesa

Reviewer 3: @tiagohermano não traduzir esse bloco ;)

```suggestion
filterRes filterUnits floodColor floodOpacity focusable fontFamily fontSize
```

xmlLang xmlSpace y y1 y2 yChannelSelector z zoomAndPan
```

> Você também pode usar atributos personalizados desde que eles estejam totalmente em letras minúsculas.

Reviewer 3: ```suggestion
Você também pode usar atributos personalizados, desde que todos eles estejam em letras minúsculas.
```
